File: 2766-find-the-prefix-common-array-of-two-arrays/find-the-prefix-common-array-of-two-arrays.java
class Solution {
    public int[] findThePrefixCommonArray(int[] A, int[] B) {
        int n = A.length ;
        int[] cntA = new int[n + 1];
        int[] cntB = new int[n + 1];

        int[] ans = new int[n];
        for(int i = 0 ; i < n ; i ++){
            if(A[i] == B[i]){
                ans[i] = (i != 0 ? ans[i - 1] : 0) + 1 ;
            }else{
                cntA[A[i]] ++ ;
                cntB[B[i]] ++ ;
                if(cntA[B[i]] != 0 && cntB[A[i]] != 0){
                    ans[i] = ans[i - 1] + 2 ;
                }else if(cntA[B[i]] != 0 || cntB[A[i]] != 0){
                    ans[i] = ans[i - 1] + 1 ;
                }else  ans[i] = (i != 0 ? ans[i - 1] : 0) ;
            }
        }
        return ans ;
    }
}

File: 2766-find-the-prefix-common-array-of-two-arrays/find-the-prefix-common-array-of-two-arrays.py
class Solution:
    def findThePrefixCommonArray(self, A: List[int], B: List[int]) -> List[int]:
        n = len(A)
        cnta, cntb = [0] * (n + 1), [0] * (n + 1)
        C = []
        for a, b in zip(A, B):
            if a == b:
                C.append(C[-1] + 1 if len(C) else 1)
            else:
                cnta[a] += 1
                cntb[b] += 1 
                if cntb[a] and cnta[b]:
                    C.append(C[-1] + 2)
                elif cnta[b] or cntb[a]:
                    C.append(C[-1] + 1)
                else:
                    C.append(C[-1] if len(C) else 0)
        
        return C